### PR TITLE
vim-patch:8.1.0825: code for autocommands is mixed with file I/O code

### PR DESCRIPTION
**Pull request by tjdevries**
Copy & Paste to new file.

Had to clean up lints & move around some declarations so they could be found by build system, but otherwise nothing.

### CODE REVIEW

**Comment by jamessan** [comment]
> Had to clean up lints

That's debatable, but thanks for doing it. :) Could've just done the copy/paste dance and then let the (not yet ported to GHA) lint job catch up after the merge.

File: src/nvim/autocmd.c [hunk 1]
```diff
@@ -0,0 +1,2089 @@
+// autocmd.c: Autocommand related functions
```
Comment (jamessan) [inline src/nvim/autocmd.c:1]: Can you run `./scripts/pvscheck.sh --patch .`?

**Comment by tjdevries** [comment]
> > Had to clean up lints
> 
> That's debatable, but thanks for doing it. :) Could've just done the copy/paste dance and then let the (not yet ported to GHA) lint job catch up after the merge.

Do we have something that auto fixes lint errors?

**Comment by jamessan** [comment]
> Do we have something that auto fixes lint errors?

No, but the clint errors are all based on the delta from the last report we generated on master.  We only look at errors that aren't already known.  In the case of copy/pasting to a new file, all of them will be new ... until the report is regenerated.

**Comment by tjdevries** [comment]
Ah, I see about the lint errors. It's OK. Most of the works was done by clang-format & clang-tidy, so after that it was just a matter of some macros and practicing vim movements ;)

**Comment by tjdevries** [comment]
@janlazo how do you usually proceed with merging after you're done (assuming everything passes here)? I believe I've ported the patch (went through function by function and made sure they were moved over). I then just had to move some type declarations to header files and unwind some includes, but it was more of a matter of "did it compile or not" :laughing: 

Thanks :)

**Comment by tjdevries** [comment]
What is this test failure?...

```
 [  FAILED  ] test/functional/ui/output_spec.lua @ 233: shell command :! powershell supports literal strings
  test/functional/ui/output_spec.lua:238: Failed to match any screen lines.
  Expected (anywhere): "
  Write%-Output %$a"
  Actual:
    |                              |
    |                              |
    |:!'Write-Output $a'           |
    |                              |
```

**Comment by janlazo** [comment]
@tjdevries I investigate if there are more related patches. For this PR, I don't think there's anything needed to be ported together with the patch 8.1.0825 because the followup patches after moving files would deal with `#ifdef`, headers, `FEAT_` macros. These followup patches are usually N/A.  After porting all related patches, I request a review from one of the maintainers. Nowadays, I merge without review if I do not have any doubts with how I ported the patches. After merging, I may update the wiki, https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim, to reflect the changes (ie. `enc_utf8`) and decisions (ie. `screedump` tests) in the merged PR.

I don't recommend merging Vim patches without review unless you're reviewed the Vim commit history enough to spot dependencies between patches. I spent a lot of time merging N/A patches across Vim versions so it should be easier to spot these dependencies now.

That test failure is unrelated. May need to add `poke_eventloop` or `sleep` to wait for powershell to finish.

**Comment by romgrk** [comment]
Overall, this LGTM, as far as I understand it's just moving code so as long as it compiles it should be fine.

One thing, it creates files `autocmd.c` and `autocmd.h` but we already have `aucmd.c` and `aucmd.h` (which I'm working on in #13247). Not sure if a rename happened in vim or something, but I guess this would be a good place to merge both?

Also I'll wait for this to be merged before continuing that PR because there will be conflicts, if someone wants to review & merge quickly it would be nice :)

**Comment by tjdevries** [comment]
Yeah, in a different commit we can change the structure more easily (but for now, didn't want to move other things into the file).

If tests pass (after the rebase I just did) I will merge.

**Comment by romgrk** [comment]
Now? Are Windows tests supposed to pass or can they be ignored?

**Comment by tjdevries** [comment]
(rebased and pushed, windows was green)

**Comment by wingpin6** [comment]
Hello. I found my way here through git bisect after merging with the latest master. It seems I can't run sudo make (install) because of an error from this PR.

The error lies in src/nvim/autocmd.h:74:11:
```
fatal error: autocmd.h.generated.h: No such file or directory
74: # include "autocmd.h.generated.h"
```
Would you know if it is an error on my side? I am running Ubuntu 20.04 and I have been on neovim nightly the past few weeks and built without issues. Please let me know if you need more info.

**Comment by bfredl** [comment]
@wingpin6 you might need to `rm -rf build/`, sometimes the cmake reconfig do not pickup changes in file structure.

**Comment by wingpin6** [comment]
Thank you for the tip, that did the trick. I'll remember that for the future.

**Pull request merged**